igvc/src/pathplanner/GraphSearch.hpp

```diff
@@ -35,16 +35,49 @@ class PathComparator
     }
 };
 
+template <class StateType, class ActionType, template <typename ...> class FrontierType>
+struct FrontierTraits
+{
+	using frontier_type = FrontierType <Path <StateType, ActionType>>;
+	using value_type = typename frontier_type::value_type;
+
+	static auto PopNextElement(frontier_type &frontier) -> value_type
+	{
+		auto next = frontier.top();
+		frontier.pop();
+		return next;
+	}
+};
+
+
+template <class StateType, class ActionType>
+struct FrontierTraits <StateType, ActionType, std::queue>
+{
+	using frontier_type = std::queue <Path<StateType, ActionType>>;
+	using value_type = typename frontier_type::value_type;
+	
+	static auto PopNextElement(frontier_type &frontier) -> value_type
+	{
+		auto next = frontier.front();
+		frontier.pop();
+		return next;
+	}
+};
+
 class GraphSearch
 {
 
 public:
-    /** Runs Depth-First graph search on the given search problem */
-    template <class StateType, class ActionType>
-    static Path<StateType, ActionType> DFS(SearchProblem<StateType, ActionType> &problem)
-    {
-        set<StateType> expanded;
-        stack< Path<StateType, ActionType> > frontier;
+	// Runs a generic search that DFS and BSF can call.
+	// Getting this to work with the current implementation of A* is possible, but difficult.
+	template <class StateType, class ActionType, template <typename ...> class FrontierType>
+	static auto GenericSearch(SearchProblem<StateType, ActionType> &problem) -> Path<StateType, ActionType>
+	{
+		using frontier_traits = FrontierTraits<StateType, ActionType, FrontierType>;
+		using frontier_type = typename frontier_traits::frontier_type;
+        
+		set<StateType> expanded;
+        frontier_type frontier;

```

Comment:
Sorry about that; I'm not sure what happened. 
I'll try to fix the spacing tomorrow.
Should I do another pull request once it's done?